feat(account-lib): skeleton code for avalanche c-chain in account-lib

shiksb opened the pull request:
Ticket: STLX-3250

OttoAllmendinger commented:
please use `feat(account-lib)` so that we can easier identify the component

smaria95 commented on modules/account-lib/src/coin/avaxc/keyPair.ts:
```diff
@@ -0,0 +1,41 @@
+
+export class KeyPair extends Secp256k1ExtendedKeyPair {
+  /**
+   * Public constructor. By default, creates a key pair with a random master seed.
+   *
+   * @param { KeyPairOptions } source Either a master seed, a private key, or a public key
+   */
+  constructor(source?: KeyPairOptions) {
+    super(source);
+    if (source) {
+      if (isPrivateKey(source)) {
+        super.recordKeysFromPrivateKey(source.prv);
+      } else if (isPublicKey(source)) {
+        super.recordKeysFromPublicKey(source.pub);
+      } else {
+        throw new Error('Invalid key pair options');
+      }
+    } else {
+      throw new Error('Invalid key pair options');
+    }
```
I think the constructor `implementation` and the `jsdoc` are not consistent within each other.

The doc says: `By default, creates a key pair with a random master seed`.
While the constructor throws an error if it's called without an argument.

I think the correct behavior is the one detailed on the jsdoc, but since this class will be implemented as part of [STLX-5078](https://bitgoinc.atlassian.net/browse/STLX-5078) maybe we can just add a comment like: `TODO(STLX-5078): Implement random seed and key pair generation`.

smaria95 commented on modules/account-lib/src/coin/avaxc/transaction.ts:
```diff
@@ -0,0 +1,69 @@
+   * @param {TransactionType} transactionType The transaction type to be set
+   */
+  setTransactionType(transactionType: TransactionType): void {
+    throw new NotImplementedError('getTransferData not implemented');
```
```suggestion
    throw new NotImplementedError('setTransactionType not implemented');
```

smaria95 commented on modules/account-lib/src/coin/avaxc/transaction.ts:
```diff
@@ -0,0 +1,69 @@
+   * and save them into the base transaction signature list.
+   */
+  loadPreviousSignatures(): void {
+    throw new NotImplementedError('getTransferData not implemented');
```
```suggestion
    throw new NotImplementedError('loadPreviousSignatures not implemented');
```

smaria95 commented on modules/account-lib/src/coin/avaxc/transactionBuilder.ts:
```diff
@@ -0,0 +1,131 @@
+   * @returns { TransactionBuilder }This transaction builder
+   */
+  signature(signature: string, keyPair: KeyPair): TransactionBuilder {
+    throw new NotImplementedError('initializeBuilder not implemented');
```
```suggestion
    throw new NotImplementedError('signature not implemented');
```

smaria95 commented on modules/account-lib/src/coin/avaxc/transactionBuilderFactory.ts:
```diff
@@ -0,0 +1,59 @@
+   * @param {any} rawTransaction - Transaction in any format
+   */
+  private validateRawTransaction(rawTransaction: any) {
+    throw new NotImplementedError('parseTransaction not implemented');
```
```suggestion
    throw new NotImplementedError('validateRawTransaction not implemented');
```

smaria95 commented on modules/account-lib/src/coin/avaxc/walletInitializationBuilder.ts:
```diff
@@ -0,0 +1,60 @@
+  /**
+   * Set one of the owners of the multisig wallet.
+   *
+   * @param {string} address The public key of the owner's account
```
I think the owners will be set as addresses and not as public keys, because that is what the contract expects to.
```suggestion
   * @param {string} address The address of the owner's account
```

CapnMigraine commented on modules/account-lib/src/coin/avaxc/transaction.ts:
```diff
@@ -0,0 +1,69 @@
+import { BaseCoin as CoinConfig } from '@bitgo/statics/dist/src/base';
```
I think we could re-use `eth`'s implementation of `Transaction`, like we did with [etc](https://github.com/BitGo/BitGoJS/blob/master/modules/account-lib/src/coin/etc/index.ts#L2)

CapnMigraine reviewed (changes requested):
Please update the `@bitgo/statics` dependency in account-lib to `6.12.0-rc.2` to access the avaxc definition.
We need to add the coin to the exports and the `coinBuilderMap` in `index.ts`
We also need an index.ts in the [coin folder](https://github.com/BitGo/BitGoJS/blob/master/modules/account-lib/src/coin/etc/index.ts) for the exports.
The classes defined for Avaxc will likely be used by the eth-like coin controller in wallet-platform, we should try to re-use as much as the `eth` implementation in account-lib as possible, so we don't have to create a different path for avaxc in wallet-platform unless we have to.

shiksb commented:
I forked this repo instead of cloning. Addressed all your comments and created a new PR for the skeleton code: https://github.com/BitGo/BitGoJS/pull/1375
